### PR TITLE
[IMM32] Rewrite ImmGetContext

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Add `ValidateHwndNoErr` and `Imm32GetContextEx` helper functions.
- Rewrite `ImmGetContext` function by using the helper functions.

### CODE REVIEW

File: dll/win32/imm32/imm.c [hunk 1]
```diff
@@ -61,6 +61,33 @@ SHAREDINFO g_SharedInfo = { NULL };
...
+    /* See if the window is cached */
+    if (hwnd == ClientInfo->CallbackWnd.hWnd)
+        return ClientInfo->CallbackWnd.pWnd;
+
+    if (!NtUserValidateHandleSecure(hwnd))
+        return NULL;
+
+    ht = g_SharedInfo.aheList; /* handle table */
+    index = (LOWORD(hwnd) - FIRST_USER_HANDLE) >> 1;
+    if (index < 0 || index >= ht->nb_handles || !ht->handles[index].type)
```
Comment (katahiromz) [inline dll/win32/imm32/imm.c:81]: @HBelusca How to check `TYPE_WINDOW`?

File: dll/win32/imm32/imm.c [hunk 2]
```diff
@@ -61,6 +61,33 @@ SHAREDINFO g_SharedInfo = { NULL };
 BYTE g_bClientRegd = FALSE;
 HANDLE g_hImm32Heap = NULL;
 
+static PWND FASTCALL ValidateHwndNoErr(HWND hwnd)
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:64]: That function should be in user32.
https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/user/user32/misc/misc.c;hb=84f7bee18f5842c63b0ad4389af8b7f49e375738#l384

Comment (learn-more) [reply dll/win32/imm32/imm.c:64]: I would expect this to be in some private header, that our components can be used.
(I have seen this symbol in multiple pdb files)

Comment (katahiromz) [reply dll/win32/imm32/imm.c:64]: This function is different in the point that it uses the handle table from `USER32`.

File: win32ss/include/ntuser.h [hunk 3]
```diff
@@ -1312,6 +1312,7 @@ C_ASSERT(sizeof(CLIENTIMC) == 0x34);
 #define CLIENTIMC_WIDE 0x1
 #define CLIENTIMC_UNKNOWN1 0x40
 #define CLIENTIMC_UNKNOWN2 0x100
+#define CLIENTIMC_UNKNOWN3 0x80
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/include/ntuser.h]: Fixed.

File: dll/win32/imm32/imm.c [hunk 4]
```diff
@@ -120,6 +147,13 @@ static inline BOOL Imm32IsCrossThreadAccess(HIMC hIMC)
     return (dwImeThreadId != dwThreadId);
 }
 
+static BOOL Imm32IsCrossProcessAccess(HWND hWnd)
+{
+    DWORD dwProcessId1 = (DWORD)(DWORD_PTR)NtCurrentTeb()->ClientId.UniqueProcess;
+    DWORD dwProcessId2 = (DWORD)NtUserQueryWindow(hWnd, QUERY_WINDOW_UNIQUE_PROCESS_ID);
+    return (dwProcessId1 != dwProcessId2);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: Fixed.

File: dll/win32/imm32/imm.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c:64]: `USER32` and `IMM32` are separate modules. And the handle table is one component in one process. So we have to use the shared info to access the handle table.

File: dll/win32/imm32/imm.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: `ht->handles[index].type != TYPE_WINDOW`
OK.

**Pull request merged**